packages/fetchai/connections/ledger/connection.py

```diff
@@ -134,8 +134,8 @@ async def receive(self, *args, **kwargs) -> Optional["Envelope"]:
             return self._handle_done_task(done_task)
 
         if len(self.receiving_tasks) == 0:
-            await self.event_new_receiving_task.wait()
             self.event_new_receiving_task.clear()
+            await self.event_new_receiving_task.wait()
```

Comment:
no need to clear after, because we wouldn't wait for the event to be set if `len(self.receiving_tasks) > 0`

Also, clear the event before, because if the list is empty we are sure we need to wait.